store/index/checkpoint.go

```diff
@@ -0,0 +1,348 @@
+package index
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-storethehash/store/types"
+)
+
+var log = logging.Logger("storethehash/index")
+
+const (
+	// checkpointInterval is how often to advance the checkpoint.
+	checkpointInterval = 2 * time.Minute
+	// checkpointRuntime is the maximum time to run checkpoint advancement if
+	// it does not find a current entry to stop on.
+	checkpointRuntime = time.Minute
+	// compactPercentThreshold is percent of the index that must be eliminated
+	// by the checkpoint for compaction to take place.
+	compactPercentThreshold = 33
+)
+
+// checkpointer is a goroutine that waits for a signal to update the index
+// checkpoint, and looks for the next checkpoint if it has not been too soon,
+// `checkpointInterval`, since the last time.  The search runs and advances the
+// checkpoint until it has been running for `checkpointRuntime` or until it
+// finds the next live index entry.  The search runs in a separate goroutine so
+// that this function can continue reading the channel used to signal updates.
+func (i *Index) checkpointer() {
+	cpPath := checkpointPath(i.file.Name())
+	cpDone := make(chan struct{})
+	var cancel context.CancelFunc
+	hasUpdate := true
+	t := time.NewTimer(checkpointInterval)
+
+	for {
+		select {
+		case _, ok := <-i.cpUpdate:
+			if !ok {
+				// Channel closed; shutting down.
+				if cancel != nil {
+					cancel()
+					<-cpDone
+				}
+				return
+			}
+			hasUpdate = true
+		case <-t.C:
+			if !hasUpdate {
+				// Nothing new, keep waiting.
+				t.Reset(checkpointInterval)
+				continue
+			}
+
+			go func() {
+				defer func() {
+					cpDone <- struct{}{}
+				}()
+				var ctx context.Context
+				ctx, cancel = context.WithTimeout(context.Background(), checkpointRuntime)
+				cp, err := i.updateCheckpoint(ctx, cpPath)
+				if err != nil {
+					log.Errorw("Failed to update checkpoint", "err", err)
+					return
+				}
+
+				// Only compact if checkpoint update finished and did not timeout.
+				if ctx.Err() == nil {
+					// See if the checkpoint reduces the size enough to make it worth it.
+					if (cp*100)/int64(i.length) >= compactPercentThreshold {
```

Comment:
we should log how many times we do the work of tracking the checkpoint but not reaching threshold in order to tune that.

Reply:
Done.  I am thinking that it might be best to always let the checkpoint update finish.  After the indexer is cought up, this should not be terrible.

I originally had the timeout so that the indexer would not devote that much CPU time to updating the checkpoint that it was taking a long time to catch up to, and to periodically record the checkpoint so that the indexer would not haved to start all over again if the process was terminated before recording the checkpoint.

Once the indexer is caught up, neither of these things seems like a problem... so I think I will change this to always finish.